src/argman.rs

```diff
@@ -0,0 +1,712 @@
+
+use std::collections::HashMap;
+use std::env;
+
+#[derive(Debug)]
+enum ArgType {
+    ArgBool,
+    ArgMultistr,
+    ArgMapStr,
+    ArgStr,
+}
+
+#[derive(Debug)]
+struct ArgumentHelp {
+    description: String,
+    arg_type: ArgType,
+    default: Option<String>,
+    default_multi: Vec<String>,
+    default_map: HashMap<String, String>,
+}
+
+pub struct ArgMan {
+    args: HashMap<String, String>,
+    args_help: HashMap<String, ArgumentHelp>,
+    args_multi: HashMap<String, Vec<String>>,
+    args_multi_map: HashMap<String, HashMap<String, String>>,
+}
+
+impl ArgMan {
+
+    pub fn new() -> ArgMan {
+        ArgMan {
+            args_help: HashMap::new(),
+            args: HashMap::new(),
+            args_multi: HashMap::new(),
+            args_multi_map: HashMap::new(),
+        }
+    }
+
+    pub fn add_arg_unset(&mut self, name: &str, description: &str) {
```

Comment:
This seems like wayyyy overkill. We have, like, two or three parameters. Instead of the complexity of a generic arg framework, why not just have an enum with the set of arguments we support and use that directly? As it is main.rs doesn't really go down in size.

Reply:
Well, I want to have an ArgMan that can be tested completely without caring about anything else in the project. Perhaps that's overkill already.
We can always remove feature we don't need from the generic thing.
For example, we can force the user to always set a default value, that can simplify ArgMan API.
Can we focus on the API for ArgMan first and then the implementation?

As clarification, reducing the size of main.rs was never the goal of this, mostly more similar to other implementations: lnd, clightning, eclair, etc. But I guess we can always move create_global_args() to a new init.rs or config.rs file for clarity, that would make main.rs smaller.

I really though you were going to complain about the multi_map feature, either because ``bin regtest.-rpcport=9999``` looks weird or becase the repod doesn't need it at all just like the args_multi feature or repeated arguments at the moment, I just need that stuff for https://github.com/jtimon/rust-lightning-bitcoinrpc/pull/2 and https://github.com/jtimon/multi-ln-demo

I can totally do a simpler version since we only need one type of get so far, look at https://github.com/jtimon/rustlnd/commit/f202fe046e45b30c306f59c9f9edb42c697b9d22 or just https://github.com/jtimon/rustlnd/commits/master for more context.

Then I will try to PR simultaneous multi-chain support, even if lightning integration only configures it for one chain, it should just at least keep working for one chain.

Anyway, please keep up with the feedback, you know I work mostly asynchronously so please don't hate me for rants lie this one.

Reply:
Right, I can't say I really care almost at all about what the format on the options looks like, mostly I just want less code in total. Reducing main.rs especially. If this ends up with a big honkin' framework that handles every possible future use of args. Its a demo app, not really a full lightning node.